### PR TITLE
Rpi 4.4.y adau1977

### PR DESCRIPTION
**Pull request by andrey2805**
The following 2 commits are result of my work on bring up of ADAU1977 ADC  on RPI 2b+.
They add basic support for the codec usage including: Device tree overlay, binding I2S bus and setting I2S mode, clock source and frequency setting according to specification.

### CODE REVIEW

**Comment by andrey2805** [comment]
Forgot to mention that the adau1977-adc is based on  HifiBerry DAC driver by Florian Meier and the guides and specs for the HW are at http://www.analog.com/en/products/audio-video/audio-ad-converters/adau1977.html#product-overview

**Comment by pelwell** [comment]
There are a number of problems with this PR:
1. Overlays belong in arch/arm/boot/dts/_overlays_.
2. Why aren't the two overlays combined into one?
3. Make the card definition in the overlay (the one with the i2s-controller property) the last fragment - that will allow it to be dynamically unloaded with crashing. 
4. Change the comment at the top of the -adc overlay.
5. Overlays need to be compiled, for which you need add them into arch/arm/boot/dts/overlays/Makefile. Please observe alphabetical order.
6. I require an entry in the README for each overlay. Copy the formatting of other entries, insert in alphabetical order and don't use TABs.
7. Your driver needs to be compiled, for which you need to add something to sound/soc/bcm/Makefile.
8. Although copying is allowed, it is polite to mention the author of any work you have copied from.
9. I don't think you need to add an of_device_id/compatible string to the I2C driver. It is a feature of the I2C subsystem that it will look for the element after the comma in a compatible string and look it up using the i2c_device_id table. In fact I think that including an explicit compatible string can stop the device type matching feature from working (the same I2C driver supports multiple devices by returning different IDs for different devices - see the use of ADAU1977 and ADAU1988 in the i2c_device_id table).
10. It is better not to modify upstream code if you can avoid it.

There may be more, but that's enough for now.

**Comment by andrey2805** [comment]
Thanks for your comments, will rework in short time.

**Comment by andrey2805** [comment]
Here are the follow-up commits.

**Comment by andrey2805** [comment]
And another minor typo...

File: arch/arm/boot/dts/overlays/README [hunk 1]
```diff
@@ -161,6 +161,13 @@ Params:
         and the other i2c baudrate parameters.
 
 
+Name:   adau1977-adc
+Info:   Overlay for activation of ADAU1977 ADC codec over I2C for conrtrol
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README]: There is a typo ("conrtrol" -> "control") and whitespace at the end of this line.

File: arch/arm/boot/dts/overlays/adau1977-adc-overlay.dts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+			};
+		};
+	};
+	
+	fragment@1 {
+        	target = <&i2c>;
+        	
+		__overlay__ {
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/adau1977-adc-overlay.dts]: Here you need:
```
                       #address-cells = <1>;
                       #size-cells = <0>;
                       status = "okay";

```

**Comment by pelwell** [comment]
The final pieces you are missing are the additions to bcmrpi_defconfig and bcm2709_defconfig to cause the codec and card driver to be built:

```
diff --git a/arch/arm/configs/bcm2709_defconfig b/arch/arm/configs/bcm2709_defconfig
index 54c8f4e..47bd1c2 100644
--- a/arch/arm/configs/bcm2709_defconfig
+++ b/arch/arm/configs/bcm2709_defconfig
@@ -861,6 +861,7 @@ CONFIG_SND_BCM2708_SOC_BOOMBERRY_DAC=m
 CONFIG_SND_BCM2708_SOC_BOOMBERRY_DIGI=m
 CONFIG_SND_BCM2708_SOC_IQAUDIO_DAC=m
 CONFIG_SND_BCM2708_SOC_RASPIDAC3=m
+CONFIG_SND_BCM2708_SOC_ADAU1977_ADC=m
 CONFIG_SND_SOC_ADAU1701=m
 CONFIG_SND_SOC_WM8804_I2C=m
 CONFIG_SND_SIMPLE_CARD=m
diff --git a/arch/arm/configs/bcmrpi_defconfig b/arch/arm/configs/bcmrpi_defconfig
index 13471cc..d66fcd4 100644
--- a/arch/arm/configs/bcmrpi_defconfig
+++ b/arch/arm/configs/bcmrpi_defconfig
@@ -853,6 +853,7 @@ CONFIG_SND_BCM2708_SOC_BOOMBERRY_DAC=m
 CONFIG_SND_BCM2708_SOC_BOOMBERRY_DIGI=m
 CONFIG_SND_BCM2708_SOC_IQAUDIO_DAC=m
 CONFIG_SND_BCM2708_SOC_RASPIDAC3=m
+CONFIG_SND_BCM2708_SOC_ADAU1977_ADC=m
 CONFIG_SND_SOC_ADAU1701=m
 CONFIG_SND_SOC_WM8804_I2C=m
 CONFIG_SND_SIMPLE_CARD=m
```

Make the changes I have suggested, then squash all the commits into one and update your PR using `git push -f`.

**Comment by andrey2805** [comment]
Done.

File: arch/arm/boot/dts/overlays/README [hunk 3]
```diff
@@ -161,6 +161,13 @@ Params:
         and the other i2c baudrate parameters.
 
 
+Name:   adau1977-adc
+Info:   Overlay for activation of ADAU1977 ADC codec over I2C for control
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README]: You still have whitespace at the end of this line.

File: sound/soc/bcm/adau1977-adc.c [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	.dai_link     = snd_rpi_adau1977_dai,
+	.num_links    = ARRAY_SIZE(snd_rpi_adau1977_dai),
+};
+
+static int snd_adau1977_adc_probe(struct platform_device *pdev)
+{
+	int ret = 0;
+
+	snd_adau1977_adc.dev = &pdev->dev;
+	dev_info(&pdev->dev,"%s:%d\n",__func__,__LINE__);
```
Comment (pelwell) [inline sound/soc/bcm/adau1977-adc.c]: Remove this line - it doesn't tell you anything useful. You will see ALSA messages for success and error.

File: sound/soc/bcm/adau1977-adc.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	    if (i2s_node) {
+		dai->cpu_dai_name = NULL;
+		dai->cpu_of_node = i2s_node;
+		dai->platform_name = NULL;
+		dai->platform_of_node = i2s_node;
+	    }
+	}
+
+	ret = snd_soc_register_card(&snd_adau1977_adc);
+	dev_info(&pdev->dev,"%s:%d\n",__func__,__LINE__);
```
Comment (pelwell) [inline sound/soc/bcm/adau1977-adc.c]: Remove this as well.

**Comment by pelwell** [comment]
Remove the trailing whitespace plus the two dev_info()s, squash and push, and I will merge it.

**Comment by andrey2805** [comment]
Done.

**Pull request merged**